debridge_node/src/modules/chain/scanning/services/SolanaReaderService.ts

```diff
@@ -1,161 +1,127 @@
         break;
       }
-      this.logger.verbose(`Transactions ${batchTransactionsHashes.length} are received`);
-      transactionsHashes.push(...batchTransactionsHashes);
-      earliestTransactionInSyncSession = batchTransactionsHashes.at(-1); // get earliest from batch
-
-      this.logger.log(`searchFrom = ${earliestTransactionInSyncSession}`);
-    } while (batchTransactionsHashes.length === this.GET_HISTORICAL_LIMIT); //getting all transactions
+    }
+  }
 
-    if (transactionsHashes.length === 0) {
-      await this.supportedChainRepository.update(
-        { chainId },
-        {
-          latestBlock: lastSolanaBlock,
-        },
-      );
+  private async createSubscription() {
+    await this.#duplex?.requests?.complete();
+    this.#submissionsFromSync = [];
+    const chain = await this.supportedChainRepository.findOne({
+      where: {
+        chainId: solanaChainId,
+      },
+    });
+    if (!chain) {
+      const message = `Chain ${solanaChainId} is not configured`;
+      this.#logger.error(message);
       return;
     }
+    this.#duplex = this.#solanaGrpcClient.getSendEvents(BigInt(chain?.latestNonce || 0), true);
 
-    transactionsHashes = transactionsHashes.reverse(); //sort for having transaction from earliest to newest
-
-    //saving transactions
-    const eventSyncingPageCount = Math.ceil(transactionsHashes.length / this.GET_EVENTS_LIMIT);
-    for (let pageNumber = 0; pageNumber < eventSyncingPageCount; pageNumber++) {
-      const skip = pageNumber * this.GET_EVENTS_LIMIT;
-      const end = Math.min((pageNumber + 1) * this.GET_EVENTS_LIMIT, transactionsHashes.length);
-      const txs = transactionsHashes.slice(skip, end);
-      this.logger.log(`Tx hashes for scan ${JSON.stringify(txs)}`);
-      const transactions = await this.solanaApiService.getEventsFromTransactions(txs);
-      const events = [];
-      transactions
-        .filter(transaction => transaction.transactionState === TransactionState.Ok)
-        .forEach(transaction => {
-          transaction.events.forEach(event => {
-            events.push({
-              ...event,
-              transactionHash: transaction.transactionHash,
-              slotNumber: transaction.slotNumber,
-            } as SolanaEvent);
-          });
-        });
-      this.logger.verbose(`Events ${events.length} from transaction are received`);
-
-      const submissions = events.map(event => {
-        try {
-          const submission = this.transformService.generateSubmissionFromSolanaEvent(event);
-          this.logger.verbose(`submission ${event.submissionId} is generated`);
-          return submission;
-        } catch (e) {
-          this.logger.error(`Can't generate submission ${event.submissionId} from solana event`);
-          this.logger.error(e);
-          throw e;
+    for await (const response of this.#duplex.responses) {
+      if (!(response.sendEventMessage !== undefined && response.sendEventMessage.oneofKind !== undefined)) {
```

Comment:
Hard to read, 
`!(sendEventMessage !== undefined && sendEventMessage.oneofKind !== undefined)` -> `if (sendEventMessage?.oneofKind === undefined) continue`